b2sdk/transfer/outbound/upload_manager.py

```diff
@@ -134,14 +138,24 @@ def _upload_part(
 
         # Retry the upload as needed
         exception_list = []
-        for _ in range(self.MAX_UPLOAD_ATTEMPTS):
-            # if another part has already had an error there's no point in
-            # uploading this part
-            if large_file_upload_state.has_error():
-                raise AlreadyFailed(large_file_upload_state.get_error_message())
-
-            try:
-                with part_upload_source.open() as part_stream:
+        with ExitStack() as stream_guard:
+            part_stream = None
+
+            def close_stream_callback(stream):
+                stream.close()
+
+            for _ in range(self.MAX_UPLOAD_ATTEMPTS):
+                # if another part has already had an error there's no point in
+                # uploading this part
+                if large_file_upload_state.has_error():
+                    raise AlreadyFailed(large_file_upload_state.get_error_message())
+
+                try:
+                    # reuse the stream in case of retry
+                    part_stream = part_stream or part_upload_source.open()
+                    # register stream closing callback only when reading is finally concluded
+                    stream_guard.callback(close_stream_callback, part_stream)
```

Comment:
Won't it add the same close callback multiple times in case of a retry? There are different sources of streams here, and some of them could consider multiple close as an error.

Also, don't you need to rewind the stream before uploading it again?

Also, if this passes, you should also add the release of the buffer on the `close` call and not on the `read` call of the `IOWrapper`. Otherwise the bug will persist.

Reply:
> Won't it add the same close callback multiple times in case of a retry? There are different sources of streams here, and some of them could consider multiple close as an error.

it will call close on the same resource indeed. Initially I had there a condition `if not stream.closed:` but calling `.close()` on closed resource does not cause any error, so I went with this version.


> Also, don't you need to rewind the stream before uploading it again?
good catch, but I believe it's done on different layer already: 

https://github.com/reef-technologies/b2-sdk-python/blob/5753d56403a388a04760a53a792caef87b429864/b2sdk/b2http.py#L240

Reply:
Did you check it on 3.11? I believe that calling `close` multiple times there causes an ugly warning... But I might have been missing something.

As for `seek` – that's true! I completely forgot about it.

Reply:
I use 3.11 as the default and didnt notice any error, but there's not reason to not play it safe, so I'll add the condition to check if it's not closed already.